Enabling check for valid nodes

papanito opened the pull request.

papanito committed:
enabling check for valid nodes

Fixes issue #206

amarts commented on cli/kubectl_kadalu/storage_add.py:
```diff
@@ -92,16 +93,13 @@ def storage_add_validation(args):
-    if args.type == "Replica2" and num_storages != 2:
+    if (args.type == "Replica1" and num_storages != 1) or
+       (args.type == "Replica2" and num_storages != 2) or
+       (args.type == "Replica3" and num_storages != 3):
```
I did the ugly coding due to some pylint warning saying more than 5 checks in a single if case is not 'advised'. But this looks clean while reading the code. I prefer this.

Two things:

`# pylint: disable=too-many-boolean-expressions` (R0916)  Above this if case.

remove the usage of `storage_mismatch` and merge the below if condition too.

Otherwise, it looks good.

amarts reviewed (commented):
Also another comment is on Commit message itself.

Let the 'title' have just 'Enabling check for valid nodes', and 'Fixes: #206' be part of the body/description of commit.

Thanks :-)

papanito committed:
remove the usage of `storage_mismatch`

Additionally:
- Fixed linting issues
- Removed code which was accidentially added by me

papanito commented:
Thanks for the review

aravindavk commented:
Thanks for the patch. 

validate_path function is called for every path/device and each time `kubectl get nodes` command is called. Instead of that, change the function as `get_kube_nodes` and call this in the beginning.

```
kube_nodes = get_kube_nodes()
```

Then while comparing,

```
for pth in paths:
    if pth.split(":")[0] not in kube_nodes:
        # Error
```

papanito commented:
@aravindavk good suggestion, will fix that

papanito committed:
Simplify solution for node validation

papanito commented:
@aravindavk now it should be fine. I do the node check for both `args.paths` and `args.device`

aravindavk commented on cli/kubectl_kadalu/storage_add.py:
```diff
@@ -92,35 +91,52 @@ def storage_add_validation(args):
+        resp = utils.execute(cmd)
+        print("The following nodes are available")
+        print(resp.stdout)
+        return resp.stdout
```
This will have nodename and other details too. Comparing node name will fail while checking.

```
nodes = []
for line in resp.stdout.split("\n"):
    nodename = line.split()[0]
    if nodename != "NAME":
        nodes.append(nodename)
return nodes
```

papanito replied:
Thanks for the suggestion, `python` is still quite new to me :-O

papanito replied:
I actually want to use kubcetl, which would be easier, bit somehow it get's me `<empty>` when running in the `cli`

```
cmd = ["kubectl", "get", "nodes", "--no-headers", "-o", "custom-columns=':metadata.name'"]
```

aravindavk commented on cli/kubectl_kadalu/storage_add.py:
```diff
@@ -92,35 +91,52 @@ def storage_add_validation(args):
                   "details in the format <node>:<path>", file=sys.stderr)
             sys.exit(1)
+        if path.split(":")[0] not in kube_nodes:
+            print("Node name does not appear to be valid")
```
good to print name of path that is failing.

aravindavk commented on cli/kubectl_kadalu/storage_add.py:
```diff
@@ -92,35 +91,52 @@ def storage_add_validation(args):
                   "details in the format <node>:<device>", file=sys.stderr)
             sys.exit(1)
+        if dev.split(":")[0] not in kube_nodes:
+            print("Node name does not appear to be valid")
```
good to print name of the dev that is failing

aravindavk reviewed (commented):
Looks good now, a few minor comments

aravindavk commented:
CI failure

```
Traceback (most recent call last):

  File "/usr/bin/kubectl-kadalu", line 11, in <module>

    load_entry_point('kubectl-kadalu===0.0.1canary', 'console_scripts', 'kubectl-kadalu')()

  File "/usr/lib/python3.5/site-packages/kubectl_kadalu-0.0.1canary-py3.5.egg/kubectl_kadalu/main.py", line 35, in main

    validation_func(args)

  File "/usr/lib/python3.5/site-packages/kubectl_kadalu-0.0.1canary-py3.5.egg/kubectl_kadalu/storage_add.py", line 105, in storage_add_validation

    kube_nodes = get_kube_nodes()

  File "/usr/lib/python3.5/site-packages/kubectl_kadalu-0.0.1canary-py3.5.egg/kubectl_kadalu/storage_add.py", line 135, in get_kube_nodes

    nodename = line.split()[0]

IndexError: list index out of range

Storage add commands failed
```

aravindavk commented on cli/kubectl_kadalu/storage_add.py:
```diff
@@ -92,35 +91,57 @@ def storage_add_validation(args):
+        print(resp.stdout)
+        nodes = []
+        for line in resp.stdout.split("\n"):
+            nodename = line.split()[0]
```
add below thing before this line to prevent the CI failure 

```
line = line.strip()
if not line:
    continue
```

papanito replied:
>  Sorry for the multiple iterations. One comment to address CI failure

No problem, sorry for the multiple updates required :-/

Ahh I see, the last line is empty

aravindavk reviewed (commented):
Looks good. Sorry for the multiple iterations. One comment to address CI failure

papanito committed:
Fix `index out of range` in `get_kube_nodes`
